File: tests/syquex/test_tiempo.py
"""
FASE 24 — Test de Tiempo (Manual 3 §12.1).

TDD: este test ES la especificación. Si _syn_tiempo_* no existen,
el test C NO compila — eso es correcto.

Manual 3 §12.1: lib/tiempo.syq — Fechas y tiempos
Comando: pytest tests/syquex/test_tiempo.py -v
Criterio: valores razonables para fecha/hora del sistema
"""
import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))

BIN_NAME = "test_tiempo.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)
TEST_SRC = os.path.join(PROJECT_ROOT, "tests", "test_tiempo.c")


def _find_gcc():
    gcc_candidate = os.path.join(PROJECT_ROOT, "toolchain_gcc12", "mingw64", "bin", "gcc.exe")
    if os.path.exists(gcc_candidate):
        return gcc_candidate
    for candidate in ("gcc", "gcc.exe"):
        try:
            subprocess.run([candidate, "--version"], capture_output=True)
            return candidate
        except FileNotFoundError:
            continue
    return gcc_candidate


@pytest.fixture(scope="module")
def exe_path():
    """Compila el test C y retorna la ruta al ejecutable."""
    if os.path.exists(BIN_ABS):
        return BIN_ABS

    gcc = _find_gcc()
    r = subprocess.run(
        [gcc, "-O2", "-I", PROJECT_ROOT, "-I.", "-o", BIN_ABS, TEST_SRC,
         "-lm", "-lpthread", "-lws2_32"],
        capture_output=True, text=True, timeout=120
    )
    if r.returncode != 0:
        pytest.fail(
            f"test_tiempo.c NO COMPILA (TDD: falta implementar §12.1).\n"
            f"Error de gcc:\n{r.stderr[-1000:]}"
        )
    assert os.path.exists(BIN_ABS), f"{BIN_NAME} no se creó"
    return BIN_ABS


def _run(exe_path):
    """Ejecuta el test C y retorna (returncode, stdout, stderr)."""
    r = subprocess.run(
        [exe_path], capture_output=True, cwd=PROJECT_ROOT,
        timeout=30, encoding="utf-8", errors="replace"
    )
    return r.returncode, r.stdout, r.stderr


class TestTiempoTimestamp:
    """§12.1 — Timestamps."""

    def test_timestamp_unix(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_tiempo falló:\n{out}\n{err}"
        assert "timestamp_unix > 1700000000" in out

    def test_timestamp_ms(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "timestamp_ms >= timestamp_unix * 1000" in out


class TestTiempoFecha:
    """§12.1 — Componentes de fecha."""

    def test_anio(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_tiempo falló:\n{out}\n{err}"
        assert "anio coincide con sistema" in out

    def test_mes(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "mes coincide con sistema" in out

    def test_dia(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "dia coincide con sistema" in out


class TestTiempoHora:
    """§12.1 — Componentes de hora."""

    def test_hora(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_tiempo falló:\n{out}\n{err}"
        assert "hora coincide con sistema" in out

    def test_minuto(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "minuto coincide con sistema" in out

    def test_segundo_rango(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "segundo en rango 0-59" in out


class TestTiempoOtros:
    """§12.1 — Día semana, día año, formato, diferencia."""

    def test_dia_semana(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_tiempo falló:\n{out}\n{err}"
        assert "dia_semana en rango 0-6" in out

    def test_dia_anio(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "dia_anio en rango 1-366" in out

    def test_fecha_formato(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "fecha_actual formato YYYY-MM-DD" in out

    def test_hora_formato(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "hora_actual formato HH:MM:SS" in out

    def test_datetime_formato(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "datetime_actual tiene espacio entre fecha y hora" in out

    def test_diferencia(self, exe_path):
        rc, out, err = _run(exe_path)
        assert "diferencia(10, 20) == 10" in out


class TestTiempoIntegracion:
    """Tests de integración completos del módulo tiempo."""

    def test_todos_los_tests_c_pasan(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_tiempo falló (rc={rc}):\n{out}\n{err}"
        assert "RESULTADO" in out, f"Sin resultados en stdout:\n{out}"
        assert "[FAIL]" not in out, f"Hay FAILs en output C:\n{out}"

File: tests/test_tiempo.c
// FASE 24 — Test de Tiempo (Manual 3 §12.1)
// TDD: este test ES la especificación. Si las funciones _syn_tiempo_*
// no existen, el test NO compila — eso es correcto.
//
// Manual 3 §12.1: lib/tiempo.syq — Fechas y tiempos
// Comando: pytest tests/syquex/test_tiempo.py -v
// Criterio: valores razonables para la fecha/hora del sistema

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include "synapse_rt_types.h"
#include "runtime/core/tiempo.h"

static int passed = 0;
static int failed = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

int main(void) {
    setbuf(stdout, NULL);
    struct tm* now = localtime(&(time_t){time(NULL)});
    int year = now->tm_year + 1900;
    int month = now->tm_mon + 1;
    int mday = now->tm_mday;
    int hour = now->tm_hour;
    int min = now->tm_min;

    // === 1. Timestamp Unix ===
    printf("=== 1. Timestamp Unix ===\n");
    int64_t ts = _syn_timestamp_unix();
    CHECK(ts > 1700000000, "timestamp_unix > 1700000000 (2023-11)");
    CHECK(ts < 2000000000, "timestamp_unix < 2000000000 (2033)");

    // === 2. Timestamp ms ===
    printf("=== 2. Timestamp ms ===\n");
    int64_t ts_ms = _syn_timestamp_ms();
    CHECK(ts_ms >= ts * 1000, "timestamp_ms >= timestamp_unix * 1000");
    CHECK(ts_ms < (ts + 2) * 1000, "timestamp_ms < (ts+2) * 1000");

    // === 3. Componentes de fecha ===
    printf("=== 3. Componentes de fecha ===\n");
    CHECK(_syn_tiempo_anio() == year, "anio coincide con sistema");
    CHECK(_syn_tiempo_mes() == month, "mes coincide con sistema");
    CHECK(_syn_tiempo_dia() == mday, "dia coincide con sistema");

    // === 4. Componentes de hora ===
    printf("=== 4. Componentes de hora ===\n");
    CHECK(_syn_tiempo_hora() == hour, "hora coincide con sistema");
    CHECK(_syn_tiempo_minuto() == min, "minuto coincide con sistema");
    CHECK(_syn_tiempo_segundo() >= 0 && _syn_tiempo_segundo() < 60,
          "segundo en rango 0-59");

    // === 5. Día de la semana ===
    printf("=== 5. Dia de semana ===\n");
    int64_t dow = _syn_tiempo_dia_semana();
    CHECK(dow >= 0 && dow <= 6, "dia_semana en rango 0-6");

    // === 6. Día del año ===
    printf("=== 6. Dia del anio ===\n");
    int64_t doy = _syn_tiempo_dia_anio();
    CHECK(doy >= 1 && doy <= 366, "dia_anio en rango 1-366");

    // === 7. Formateo fecha ===
    printf("=== 7. Formateo fecha ===\n");
    CadenaSegura fecha = _syn_tiempo_fecha_actual();
    CHECK(fecha.longitud == 10, "fecha_actual longitud == 10 (YYYY-MM-DD)");
    CHECK(fecha.datos[4] == '-' && fecha.datos[7] == '-',
          "fecha_actual formato YYYY-MM-DD");

    // === 8. Formateo hora ===
    printf("=== 8. Formateo hora ===\n");
    CadenaSegura hora_s = _syn_tiempo_hora_actual();
    CHECK(hora_s.longitud == 8, "hora_actual longitud == 8 (HH:MM:SS)");
    CHECK(hora_s.datos[2] == ':' && hora_s.datos[5] == ':',
          "hora_actual formato HH:MM:SS");

    // === 9. Formateo datetime ===
    printf("=== 9. Formateo datetime ===\n");
    CadenaSegura dt = _syn_tiempo_datetime_actual();
    CHECK(dt.longitud == 19, "datetime_actual longitud == 19 (YYYY-MM-DD HH:MM:SS)");
    CHECK(dt.datos[10] == ' ', "datetime_actual tiene espacio entre fecha y hora");

    // === 10. Diferencia ===
    printf("=== 10. Diferencia ===\n");
    CHECK(_syn_tiempo_diferencia_segundos(ts + 10, ts + 20) == 10,
          "diferencia(10, 20) == 10");
    CHECK(_syn_tiempo_diferencia_segundos(ts, ts) == 0,
          "diferencia(ts, ts) == 0");
    CHECK(_syn_tiempo_diferencia_segundos(ts + 5, ts) == -5,
          "diferencia(5, 0) == -5");

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}
